src/Microsoft.OData.Core/ODataNotificationWriter.cs

```diff
@@ -324,13 +331,40 @@ protected override void Dispose(bool disposing)
                 if (this.listener != null)
                 {
                     // Tell the listener that the stream is being disposed.
-                    this.listener.StreamDisposed();
+                    if (synchronous)
+                    {
+                        this.listener.StreamDisposed();

```

Comment:
Maybe call `this.listener?.StreamDispose()` so you can get rid of the `this.listener != null` block?

Reply:
@habbes There are two statements in that block:
```csharp
this.listener.StreamDisposed();
this.listener = null;
```
The second statement is important especially because if someone did something like this:
```csharp
notificationWriter.Dispose();
notificationWriter.Dispose();
// Or
await notificationWriter.DisposeAsync();
await notificationWriter.DisposeAsync();
```
Checking if listener is null guarantees we'll only ever call `StreamDisposed` or `StreamDisposedAsync` once.
That can also be guaranteed by the change you suggested in this way:
```csharp
this.listener?.StreamDisposed();
this.listener = null;
```
But I'm not particularly convinced that's a significant improvement from:
```csharp
if (this.listener != null)
{
    this.listener.StreamDisposed();
    this.listener = null;
}
```